File: constraints/dataAnalysis/stellarMassFunctions_PRIMUS_z0_1/massErrors.pl
#!/usr/bin/perl
use strict;
use Astro::FITS::CFITSIO qw( :constants );
use PDL;
use PDL::NiceSlice;
use PDL::LinearAlgebra;
use PDL::IO::HDF5;
use Carp;
Astro::FITS::CFITSIO::PerlyUnpacking(0);

# Fit polynomials to the mass and redshift dependence of stellar mass errors in the PRIMUS survey fields.
# Andrew Benson (22-May-2014)

# Get field solid angles.
my $solidAngleFile = new PDL::IO::HDF5("constraints/dataAnalysis/stellarMassFunctions_PRIMUS_z0_1/solidAngles.hdf5");
my $solidAngles    = $solidAngleFile->dataset("solidAngle")->get();
my $weights        = $solidAngles/sum($solidAngles);
# Define files for each field.
my @fileNames = 
    (
     "cosmos_benson.fits"    ,  
     "xmm_swire_benson.fits" ,
     "cfhtls_xmm_benson.fits",  
     "cdfs_benson.fits"      ,  
     "es1_benson.fits"
    );
# Iterate over files.
my $iField = 0;
open(my $fortranHndl,">constraints/dataAnalysis/stellarMassFunctions_PRIMUS_z0_1/massErrors.F90");
open(my $latexHndl  ,">constraints/dataAnalysis/stellarMassFunctions_PRIMUS_z0_1/massErrors.tex");
for my $fileName ( @fileNames ) {
    ++$iField;
    # Open the file and select the relevant HDU.
    my $status;
    my $filePath = "constraints/dataAnalysis/stellarMassFunctions_PRIMUS_z0_1/".$fileName;
    die("massErrors.pl: this script requires the PRIMUS data tables be available")
	unless ( -e $filePath );
    my $file = Astro::FITS::CFITSIO::open_file($filePath,READONLY,$status);
    $file->movabs_hdu(2,ANY_HDU,$status);
    # Extract the number of rows in the table.
    $file->get_num_rows(my $rowCount,$status);
    # Extract the data from the required columns.
    my $cols;
    foreach ( "ZPRIMUS", "MASS_50", "MASS_ERR" ) {
	$file->get_colnum(0,$_,my $dataColumn,$status);
	$cols->{$_} = zeroes($rowCount)->double;
	$file->read_col(Astro::FITS::CFITSIO::TDOUBLE(),$dataColumn,1,1,$rowCount,0,${$cols->{$_}->get_dataref},undef,$status);
	$cols->{$_}->upd_data;
    }
    # Close the file.
    $file->close_file($status);
    # Construct matrices required to optimize the polynomial coefficients.
    my @sequence = # Sequence of exponents on mass and redshift.
	(
	 [0,0],
	 [1,0],
	 [2,0],
	 [0,1],
	 [0,2],
	 [1,1] 
	);
    my $b = pdl zeroes(scalar(@sequence)                  );
    my $a = pdl zeroes(scalar(@sequence),scalar(@sequence));
    for(my $i=0;$i<scalar(@sequence);++$i) {
	my $miExponent = ${$sequence[$i]}[0];
	my $ziExponent = ${$sequence[$i]}[1];
	$b->(($i)) .= sum($cols->{'MASS_ERR'}*$cols->{'MASS_50'}**$miExponent*$cols->{'ZPRIMUS'}**$ziExponent);
	for(my $j=0;$j<scalar(@sequence);++$j) {
	    my $mjExponent = ${$sequence[$j]}[0];
	    my $zjExponent = ${$sequence[$j]}[1];
	    $a->(($i),($j)) .= 
		sum(
		    $cols->{'MASS_50'}**$miExponent*$cols->{'ZPRIMUS'}**$ziExponent
		    *
		    $cols->{'MASS_50'}**$mjExponent*$cols->{'ZPRIMUS'}**$zjExponent
		);
	}
    }
    # Solve for polynomial coefficients.
    my $x = msolve($a,transpose($b));
    # Output results.
    (my $fieldName = uc($fileName)) =~ s/_BENSON\.FITS//;
    $fieldName =~ s/_/ /g;
    print $fortranHndl "   ! Error fit for field: ".$fieldName."\n";
    print $fortranHndl "   error(".$iField.")=                                             &\n";
    for(my $i=0;$i<scalar(@sequence);++$i) {
	my $miExponent   = ${$sequence[$i]}[0];
	my $ziExponent   = ${$sequence[$i]}[1];
	my $coefficient  = $x->((0),($i));
	(my $pCoefficient = sprintf("%16.10e",$coefficient)) =~ tr/e/d/;
	my $prefix      = "";
	$prefix         = "+"
	    if ( $coefficient >= 0.0 );
	my $massTerm = "*logMass**" .$miExponent;
	$massTerm    = "*logMass   "
	    if ( $miExponent == 1 );
	$massTerm    = "           "
	    if ( $miExponent == 0 );
	my $redshiftTerm = "*redshift**" .$ziExponent;
	$redshiftTerm    = "*redshift   "
	    if ( $ziExponent == 1 );
	$redshiftTerm    = "            "
	    if ( $ziExponent == 0 );
	print $fortranHndl "      &         ".$prefix.$pCoefficient.$massTerm.$redshiftTerm;
	print $fortranHndl " & "
	    unless ( $i == scalar(@sequence)-1 );
	print $fortranHndl "\n";
    }

    print $latexHndl $fieldName;
    for(my $i=0;$i<scalar(@sequence);++$i) {
	my $coefficient  = $x->((0),($i));
	my $pCoefficient = sprintf("%6.4f",$coefficient);
	my $prefix      = "";
	$prefix         = "+"
	    if ( $coefficient >= 0.0 );
	print $latexHndl " & ".$prefix.$pCoefficient;
    }
    print $latexHndl " \\\\\n";
}
close($fortranHndl);
close($latexHndl  );

exit;

File: constraints/dataAnalysis/stellarMassFunctions_PRIMUS_z0_1/massErrors.py
#!/usr/bin/env python3
# Fit polynomials to the mass and redshift dependence of stellar mass errors in the PRIMUS survey fields.
# Andrew Benson (22-May-2014)

import os
import sys
import numpy as np
import h5py
from astropy.io import fits

execPath = os.environ.get('GALACTICUS_EXEC_PATH', '')

# Get field solid angles.
with h5py.File(dataPath + '/static/surveyGeometry/PRIMUS/solidAngles.hdf5', 'r') as f:
    solidAngles = f['solidAngle'][:]
weights = solidAngles / solidAngles.sum()

# Define files for each field.
fileNames = [
    'cosmos.fits',
    'xmm_swire.fits',
    'cfhtls_xmm.fits',
    'cdfs.fits',
    'es1.fits',
]

fortranLines = []
latexLines   = []

# Sequence of exponents on mass and redshift.
sequence = [
    (0, 0),
    (1, 0),
    (2, 0),
    (0, 1),
    (0, 2),
    (1, 1),
]

for iField, fileName in enumerate(fileNames):
    fieldNum  = iField + 1
    filePath  = execPath + '/constraints/dataAnalysis/stellarMassFunctions_PRIMUS_z0_1/' + fileName
    if not os.path.exists(filePath):
        raise FileNotFoundError(
            f'massErrors.py: this script requires the PRIMUS data tables be available ({filePath})'
        )

    with fits.open(filePath) as hdul:
        table     = hdul[1].data
        zPrimus   = table['ZPRIMUS' ].astype(float)
        mass50    = table['MASS_50' ].astype(float)
        massErr   = table['MASS_ERR'].astype(float)

    # Construct matrices required to optimise the polynomial coefficients.
    nSeq = len(sequence)
    b = np.zeros(nSeq)
    a = np.zeros((nSeq, nSeq))
    for i, (mi, zi) in enumerate(sequence):
        b[i] = np.sum(massErr * mass50**mi * zPrimus**zi)
        for j, (mj, zj) in enumerate(sequence):
            a[i, j] = np.sum(mass50**mi * zPrimus**zi * mass50**mj * zPrimus**zj)

    # Solve for polynomial coefficients.
    x = np.linalg.solve(a, b)

    # Format field name.
    fieldName = fileName.upper().replace('_BENSON.FITS', '').replace('_', ' ')

    fortranLines.append(f'   ! Error fit for field: {fieldName}\n')
    fortranLines.append(f'   error({fieldNum})=                                             &\n')
    for i, (mi, zi) in enumerate(sequence):
        coeff        = x[i]
        pCoeff       = f'{coeff:16.10e}'.replace('e', 'd')
        prefix       = '+' if coeff >= 0.0 else ''
        massTerm     = f'*logMass**{mi}' if mi > 1 else ('*logMass   ' if mi == 1 else '           ')
        redshiftTerm = f'*redshift**{zi}' if zi > 1 else ('*redshift   ' if zi == 1 else '            ')
        fortranLines.append(f'      &         {prefix}{pCoeff}{massTerm}{redshiftTerm}')
        if i < len(sequence) - 1:
            fortranLines.append(' & \n')
        else:
            fortranLines.append('\n')

    latexLine = fieldName
    for i in range(nSeq):
        coeff  = x[i]
        pCoeff = f'{coeff:6.4f}'
        prefix = '+' if coeff >= 0.0 else ''
        latexLine += f' & {prefix}{pCoeff}'
    latexLines.append(latexLine + ' \\\\\n')

with open('constraints/dataAnalysis/stellarMassFunctions_PRIMUS_z0_1/massErrors.F90', 'w') as f:
    f.writelines(fortranLines)

with open('constraints/dataAnalysis/stellarMassFunctions_PRIMUS_z0_1/massErrors.tex', 'w') as f:
    f.writelines(latexLines)
